rust/connlib/shared/src/callbacks.rs

```diff
@@ -39,6 +42,169 @@ impl From<Ipv6Network> for Cidrv6 {
     }
 }
 
+#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum Status {
+    Unknown,
+    Online,
+    Offline,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Hash)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum ResourceDescription {
+    Dns(ResourceDescriptionDns),
+    Cidr(ResourceDescriptionCidr),
+}
+
+impl ResourceDescription {
+    pub fn with_status(
+        r: crate::messages::client::ResourceDescription,
+        status: Status,
+    ) -> ResourceDescription {
+        match r {
+            crate::messages::client::ResourceDescription::Dns(r) => {
+                ResourceDescription::Dns(ResourceDescriptionDns::with_status(r, status))
+            }
+            crate::messages::client::ResourceDescription::Cidr(r) => {
+                ResourceDescription::Cidr(ResourceDescriptionCidr::with_status(r, status))
+            }
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        match self {
+            ResourceDescription::Dns(r) => &r.name,
+            ResourceDescription::Cidr(r) => &r.name,
+        }
+    }
+
+    pub fn status(&self) -> Status {
+        match self {
+            ResourceDescription::Dns(r) => r.status,
+            ResourceDescription::Cidr(r) => r.status,
+        }
+    }
+
+    pub fn id(&self) -> ResourceId {
+        match self {
+            ResourceDescription::Dns(r) => r.id,
+            ResourceDescription::Cidr(r) => r.id,
+        }
+    }
```

Comment:
Hm we can't flip ResourceDescription into a struct that shares these fields, because it would break compatibility with older code, right? Or if this type of `ResourceDescription` is only used within the Client, between connlib and the GUI, it could be done?

e.g.
```rust
struct ResourceDescription {
    id: ResourceId,
    name: String,
    status: Status,
    other: ResourceType,
}

enum ResourceType {
    Cidr,
    Dns,
}
```

Reply:
But the other ones look like they have the same type, since there's no `into`

Reply:
yeah, but some clients uses the `address`, maybe we can convert it to a `String` but I'm not sure having a single type will simplify things, specially since these are used only for serialization and we don't need to access the fields.